examples/mapreduce/mapreduce.cc

```diff
@@ -0,0 +1,270 @@
+        int size = mk.size();;
+        ar & size;
+        typename MapKey<T>::const_iterator it = mk.begin();
+        while (size--) {
+          ar & it->first;
+          ar & it->second;
+          it++;
+        }
+      }
+    };
+
+    template <class Archive, typename T>
+    struct ArchiveLoadImpl<Archive, MapKey<T>> {
+      static inline void load(const Archive& ar, MapKey<T>& mk) {
+        int size;
+        ar & size;
+        while (size--) {
+          std::string s;
+          T v;
+          ar & s;
+          ar & v;
+          mk.insert(std::make_pair(s, v));
+        }
+      }
+    };
+  }
+}
+
+template <typename T>
+std::ostream& operator<<(std::ostream& s, const Key<T>& key) {
+  s << "Key((" << key.first.first << "," << key.first.second << "), " << key.second << ")";
+  return s;
+}
+
+template<typename T>
+auto make_reader(Edge<Key<T>, std::string>& mapEdge)
+{
+  auto f = [](const Key<T>& filename, std::tuple<Out<Key<T>,std::string>>& out) {
+    //check if file exists
+    std::ifstream fin(filename.first.first);
+    std::filesystem::path p{filename.first.first};
+
+    std::cout << "The size of " << p.u8string() << " is "
+              << std::filesystem::file_size(p) << " bytes.\n";
+
+    if (!fin) {
+      std::cout << "File not found : " << fin << std::endl;
+      ttg_abort();
+    }
+
+    //Read the file in chunks and send it to a mapper.
+    std::string buffer; //reads only the first BLOCK_SIZE bytes
+    buffer.resize(BLOCK_SIZE);
+    int first = 0;
+    int chunkID = 0;
+
+    while(!fin.eof()) {
+      char * b = const_cast< char * >( buffer.c_str() );
+      fin.read(b + first, BLOCK_SIZE );
+      std::streamsize s = first + fin.gcount();
+      buffer.resize(s);
+      //Special handling to avoid splitting words between chunks.
+      if (s > 0) {
+        auto last = buffer.find_last_of(" \t\n");
+        first = s - last - 1;
+        std::string tmp;
+        if (fin) {
+          tmp.resize(BLOCK_SIZE + first);
+          if (first > 0) tmp.replace(0, first, buffer, last + 1, first);
+        }
+        buffer.resize(last);
+        //std::cout << buffer << std::endl;
+        send<0>(std::make_pair(std::make_pair(filename.first.first, chunkID), 0), buffer, out);
+        buffer = tmp;
+        chunkID++;
+      }
+    }
+  };
+
+  return make_tt<Key<T>>(f, edges(), edges(mapEdge), "reader", {}, {"mapEdge"});
+}
+
+template<typename T>
+void mapper(std::string chunk, MapKey<T>& resultMap) {
+  //Prepare the string by removing all punctuation marks
+  chunk.erase(std::remove_if(chunk.begin(), chunk.end(),
+          []( auto const& c ) -> bool { return ispunct(c); } ), chunk.end());
+  std::istringstream ss(chunk);
+  std::string word;
+
+  while (ss >> word)
+  {
+    std::transform(word.begin(), word.end(), word.begin(), ::tolower);
+    //std::cout << "Mapped " << word << std::endl;
+    resultMap.insert(std::make_pair(word, 1));

```

Comment:
What if the chunk contains the same word twice? And why use a multimap in the first place?

Reply:
I used multimap to be able to hold duplicate keys and have them sorted which makes counting easier in this example.